crates/mun_compiler/src/diagnostics_snippets.rs

```diff
@@ -0,0 +1,372 @@
+use mun_hir::diagnostics::Diagnostic as HirDiagnostic;
+use mun_hir::{HirDatabase, HirDisplay};
+use mun_syntax::{
+    ast, AstNode, Parse, SourceFile, SyntaxError, SyntaxKind, SyntaxNodePtr, TextRange,
+};
+
+use std::sync::Arc;
+
+use mun_hir::line_index::LineIndex;
+
+use crate::annotate_snippets_builders::{AnnotationBuilder, SliceBuilder, SnippetBuilder};
+
+use annotate_snippets::snippet::{AnnotationType, Snippet};
+
+fn text_range_to_tuple(text_range: TextRange) -> (usize, usize) {
+    (text_range.start().to_usize(), text_range.end().to_usize())
+}
+
+fn syntax_node_ptr_location(
+    syntax_node_ptr: SyntaxNodePtr,
+    parse: &Parse<SourceFile>,
+) -> TextRange {
+    match syntax_node_ptr.kind() {
+        SyntaxKind::FUNCTION_DEF => {
+            ast::FunctionDef::cast(syntax_node_ptr.to_node(parse.tree().syntax()))
+                .map(|f| f.signature_range())
+                .unwrap_or_else(|| syntax_node_ptr.range())
+        }
+        SyntaxKind::STRUCT_DEF => {
+            ast::StructDef::cast(syntax_node_ptr.to_node(parse.tree().syntax()))
+                .map(|s| s.signature_range())
+                .unwrap_or_else(|| syntax_node_ptr.range())
+        }
+        _ => syntax_node_ptr.range(),
+    }
+}
+
+pub(crate) fn syntax_error(
+    syntax_error: &SyntaxError,
+    _: &impl HirDatabase,
+    _: &Parse<SourceFile>,
+    relative_file_path: &str,
+    source_code: &str,
+    source_code_len: usize,
+    line_index: &Arc<LineIndex>,
+) -> Snippet {
+    let mut snippet = SnippetBuilder::new()
+        .title(
+            AnnotationBuilder::new(AnnotationType::Error)
+                .label("syntax error".to_string())
```

Comment:
I'd prefer if the `label` function would accept `&str` instead of `String` only. That makes the API easier to use. Internally you can convert it to `String`. This one is also neat: https://doc.rust-lang.org/std/string/trait.ToString.html